cmd/strelaypoolsrv/stats.go

```diff
@@ -210,4 +218,34 @@ func deleteMetrics(host string) {
 	relayGoRoutines.DeleteLabelValues(host)
 	relaySessionRate.DeleteLabelValues(host)
 	relayGlobalRate.DeleteLabelValues(host)
+	delete(lastStats, host)
+}
+
+// Due to some unexplainable behaviour, some of the numbers sometimes travel slightly backwards (by less than 1%)
+// This happens between scrapes, which is 30s, so this can't be a race.
+// This causes prometheus to assume a "rate reset", hence causes phenomenal spikes.
+// One of the number that moves backwards is BytesProxied, which atomically increments a counter with numeric value
+// returned by net.Conn.Read(). I don't think that can return a negative value, so I have no idea what's going on.
+func mergeStats(new stats, old stats) stats {
+	new.UptimeSeconds = mergeValue(new.UptimeSeconds, old.UptimeSeconds)
+	new.PendingSessionKeys = mergeValue(new.PendingSessionKeys, old.PendingSessionKeys)
+	new.ActiveSessions = mergeValue(new.ActiveSessions, old.ActiveSessions)
+	new.Connections = mergeValue(new.Connections, old.Connections)
+	new.Proxies = mergeValue(new.Proxies, old.Proxies)
+	new.BytesProxied = mergeValue(new.BytesProxied, old.BytesProxied)
+	new.GoRoutines = mergeValue(new.GoRoutines, old.GoRoutines)
+	new.Options.SessionRate = mergeValue(new.Options.SessionRate, old.Options.SessionRate)
+	new.Options.GlobalRate = mergeValue(new.Options.GlobalRate, old.Options.GlobalRate)
+	return new
+}
+
+func mergeValue(new, old int) int {
+	if new >= old || old == 0 || new == 0 {
+		return new
+	}
+	delta := (float64(old-new) * 100) / float64(new)
+	if delta > 1.0 {
```

Comment:
Does this intentionally return `new` if `old` is 1% *larger* than `new` or is my brain failing? And what's the 1% thing anyway? From the comment on `mergeStats` I'd expect a simple `if new < old { return old } return new`.

Reply:
We still need to capture actual rate resets, if I restart the relay server, the number has to go down.

Reply:
Right. and the sign business, i.e. `old-new` vs `new-old`? To me it looks like `new` is returned for `old > 1.01 * new` and `old` in all other cases - which would be wrong.

Reply:
```
if new > old {
  return new
}
if new > 0.99 * old {
  return old // slight backwards movement
}
return new // counter reset
}
```
?

Reply:
![image](https://user-images.githubusercontent.com/1144861/43044806-3014aae4-8da5-11e8-800e-29f0768403ac.png)

Reply:
```
In [1]: def merge(new, old):
   ...:     delta = (float(old-new)*100) / float(new)
   ...:     if delta > 1.0:
   ...:         return new
   ...:     return old
   ...: 

In [2]: merge(1001, 1000)
Out[2]: 1000
```

Reply:
:(